meson: restore tools/meson-vcs-tag.sh

keszybz opened the pull request:
This conceptually reverts e95acdfe1d3a790e18617bb992a712b34f41800d, but the actual contents of the script are taken from the command invocation in meson with all the updates that happened in the meantime.

The only functional change is that I replaced () by {}: this avoids one subprocess spawn. People were worried about the cost of vcs_tag(), and this microoptimization may help a bit. I measured the speed on machine, and noop rebuilds are still about ~120 ms.

The script is needed for in-place rpm builds, see README.build-in-place.md [1], where it is invoked from the spec file to determine the project version.

[1] https://src.fedoraproject.org/rpms/systemd/blob/rawhide/f/README.build-in-place.md

medhefgo commented:
**~~Star~~Version Wars: git-describe strikes back**

I believe this fails to create the correct output if `version_tag` is passed while building from a git tree (arch does that).

yuwata commented on meson.build:
```diff
@@ -1808,24 +1808,11 @@ xml_helper_py = find_program('tools/xml_helper.py')
-        version_cmd = '''echo "$2" '''
-endif
-
+version_cmd = project_source_root + '/tools/meson-vcs-tag.sh'
```
Is it possible to use `find_program()`?

keszybz commented:
> I believe this fails to create the correct output if `version_tag` is passed while building from a git tree (arch does that).

Oh, indeed. I changed the program to take `version_tag` and the fallback project version as separate arguments. I think it's all fixed now.

> Is it possible to use find_program()?

Yes, updated. 

> (Honestly, I'd rather see rpm learn how to extract the version info from the build artifacts.)

It's something of a chicken-and-egg problem. Currently the version is used to determine what the sources are, etc. We *could* determine the version from the sources, but then we'd need to specify what the sources are in a separate way. I think that actually specifying the version explictly is quite good. If we were extracting it from some other place, poeple would often upload the wrong sources and not see that they're building something strange for a long time.

keszybz commented:
> > Is it possible to use find_program()?
> 
> Yes, updated.

Oh, computer says no: `meson.build:1812: WARNING: Project targets '>= 0.60.0' but uses feature introduced in '0.63.0': vcs_tag with custom_tgt, external_program, or exe as the first argument.`

I'll revert it back to a string argument.

yuwata commented on tools/meson-vcs-tag.sh:
```diff
@@ -0,0 +1,25 @@
+    # and we fall back to project_version-commitid instead.
+    {
+        [ -e .git ] && git -C "$dir" describe --abbrev=7 --dirty=^ 2>/dev/null ||
+                echo "$fallback-$(git -C "$1" describe --always --abbrev=7)"
```
`$1` -> `$dir`

Is this work even if .git directory does not exist?

keszybz replied:
It did not. I reworked the conditional.

yuwata commented on tools/meson-vcs-tag.sh:
```diff
@@ -0,0 +1,25 @@
+    # If the working tree has no tags (CI builds), the first git-describe will fail
+    # and we fall back to project_version-commitid instead.
+    {
+        [ -e .git ] && git -C "$dir" describe --abbrev=7 --dirty=^ 2>/dev/null ||
```
`[ -e .git ]` -> `[ -e ${dir}/.git ]` ??

keszybz replied:
It seems that meson does chdir to project root, so this does work. But yeah, it's better to not rely on this. I"ll change it as suggested.

keszybz commented:
Thank you for the review. I think it should cover all cases now:
- out-of-the-box:  "254-985-g17c7a31"
- `meson configure build -Dversion-tag=255-pre`: "255-pre"
- `cd /var/tmp && meson setup $OLDPWD build2 && ninja -C build2 version.h && tail -n1 build2/version.h`: "254-985-g17c7a31"
- `mv .git{,.no} && touch meson.build && ninja -C build -v version.h &&  tail -n1 build2/version.h`:  "254"

yuwata commented on meson.build:
```diff
@@ -1808,26 +1808,13 @@ xml_helper_py = find_program('tools/xml_helper.py')
-        # TODO: Use 'sh' variable with meson >= 0.63.0
-        command: ['sh', '-euc', version_cmd,
-                  '_',
+        command: [project_source_root / 'tools/meson-vcs-tag.sh',
```
Maybe, better to comment we should use `find_program()` with meson >= 0.63.0.

keszybz replied:
I don't think it's better. The program is ever going to be used in one place only, so `find_program()` would make the output more verbose are require one more line of code. So I don't think we'd want to switch even if there was no problem with older versions.

yuwata reviewed (approved):
LGTM.

keszybz commented:
jammy-ppc64el / upstream failed with: 4828s TEST-69-SHUTDOWN:                   FAIL     (  6 s)
This looks suspicious, but is not related to the changes here.
